checks/raw/testdata/Directory.Pinned.packages.props

```diff
@@ -0,0 +1,8 @@
+<Project>
+  <PropertyGroup>
+    <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageVersion Include="SomePackage" Version="4.4.5" />
```

Comment:
Package versions are not pinned just because a version is declared.  NuGet just uses this version specified when it comes across a `<PackageReference Include="SomePackage" />` and emits an error if you specified a version in a project.

To truly "pin" a transitive package version that you do not have a direct `<PackageReference />` for, you have to enable transitive pinning: https://learn.microsoft.com/nuget/consume-packages/Central-Package-Management#transitive-pinning

Reply:
Nevermind, I was wrong what "pinning" meant in this context.

Reply:
[We define "pinned"](https://github.com/ossf/scorecard/blob/28db9a99a4c2e43e1f90a5afed6753e6ef2c36d3/docs/checks.md#pinned-dependencies) in a reproducibility sense: "a dependency that is explicitly set to a specific hash instead of
allowing a mutable version or range of versions". Which Avishay added support for via `nuget --locked-mode`, or the `RestoreLockedMode` property .

The NuGet claim in the PR description is that:

> Today, scorecard only supported nuget pinning dependencies with lockfiles. However, nuget provides package-manager-level assurance that when enabling CPM AND declaring specific versions for all direct dependencies the same level of pinning is provided to the end-user.

Which to my limited understanding, is based on the [package immutability nuget.org provides](https://github.com/NuGet/Home/wiki/Package-Immutability). But I'm not sure that fully covers all scenarios mentioned in the [lockfile blog post](https://devblogs.microsoft.com/nuget/enable-repeatable-package-restores-using-a-lock-file/#why-use-a-lock-file).

Reply:
@spencerschrock, could you please elaborate which scenarios mentioned in the [lockfile blog post](https://devblogs.microsoft.com/nuget/enable-repeatable-package-restores-using-a-lock-file/#why-use-a-lock-file) you are referring to as missing from this PR:

- nuget.config mismatch: is covered with CPM, since CPM suggests that packages are managed centrally and the user gets [warnings](https://learn.microsoft.com/en-us/nuget/consume-packages/Central-Package-Management#warning-when-using-multiple-package-sources) when using multiple package sources
- Intermediate versions and Floating versions - This PR checks for specifying specific versions and will deem floating versions as "unpinned"
- Package deletion is granted by [nuget](https://learn.microsoft.com/en-us/nuget/nuget-org/policies/deleting-packages)
- Package content mismatch is granted by [Package Immutability/signing](https://github.com/NuGet/Home/wiki/Package-Immutability)

Reply:
I've been out on the office on personal time, will take a look in the next day or two.

Reply:
> Intermediate versions and Floating versions - This PR checks for specifying specific versions and will deem floating versions as "unpinned"

I believe my comment was for this example:
* Day 1: If you specified `<PackageReference Include="My.Sample.Lib" Version="4.0.0"/>` but the versions available on the NuGet repositories were 4.1.0, 4.2.0 and 4.3.0, NuGet resolves to 4.1.0 as it the nearest minimum version.
* Day 2: Version 4.0.0 gets published. NuGet now restores version 4.0.0 as it is an exact match.

Reply:
I’m a bit unclear on the terminology being used here, and I’d appreciate some clarification. In the example from `checks.md`, it says:

> If your project is producing an application, declare all your dependencies with **specific versions** in your package format file (e.g. package.json for npm, requirements.txt for python, **packages.config for nuget**)

My understanding has always been that specifying exact versions is what we consider a pinned dependency check, and this is how we refer to it. Pinning provides consistency, and we typically **improve reproducibility further** with lock files for added determinism.

The example given, where the package manager resolves to the nearest version when an exact match isn’t available, doesn’t seem that compelling since other package managers handle similar cases. In practice, people pin versions that actually exist, and our registry follows an immutability policy, ensuring versions aren’t removed and that these dependencies remain reliable.

It might be helpful to create documentation that explains the different levels/scenarios of "pinned" dependency checks, with clear examples for each scenario. For instance, detailing how an exact version match qualifies as "pinned," a lock file represents "locked," and so forth, to ensure clarity across all cases.

Reply:
> * Day 2: Version 4.0.0 gets published. NuGet now restores version 4.0.0 as it is an exact match.

Typically, we see folks publishing higher versions of their packages as time goes by, not lower versions.

Reply:
> It might be helpful to create documentation that explains the different levels/scenarios of "pinned" dependency checks, with clear examples for each scenario. For instance, detailing how an exact version match qualifies as "pinned," a lock file represents "locked," and so forth, to ensure clarity across all cases.

Most of our documentation uses "pinned" to refer to lock files, so we can certainly try to clarify "pinned" vs "locked".

I did some digging into the history of the check, as well as checking with relevant contributors.   It started off as [`Frozen-Deps`, which looked for lockfiles.](https://github.com/ossf/scorecard/blob/68e801f1f80a3cc3f36ff12cca4114b07e839a47/checks/checks.md#frozen-deps), before being renamed `Pinned-Dependencies`. Most of the check is focused on lockfiles and immutable hashes (OCI or GitHub), but there is some pinning recognized for package managers with immutable versions (Go and Nuget).

> In practice, people pin versions that actually exist

> Typically, we see folks publishing higher versions of their packages as time goes by, not lower versions.

I recognize this is an edge case we probably don't need to worry about. I was just highlighting one difference of lockfiles and version pinning, as I clarified the semantics of the check.

Reply:
> > It might be helpful to create documentation that explains the different levels/scenarios of "pinned" dependency checks, with clear examples for each scenario. For instance, detailing how an exact version match qualifies as "pinned," a lock file represents "locked," and so forth, to ensure clarity across all cases.
> 
> Most of our documentation uses "pinned" to refer to lock files, so we can certainly try to clarify "pinned" vs "locked".
> 
> I did some digging into the history of the check, as well as checking with relevant contributors. It started off as [`Frozen-Deps`, which looked for lockfiles.](https://github.com/ossf/scorecard/blob/68e801f1f80a3cc3f36ff12cca4114b07e839a47/checks/checks.md#frozen-deps), before being renamed `Pinned-Dependencies`. Most of the check is focused on lockfiles and immutable hashes (OCI or GitHub), but there is some pinning recognized for package managers with immutable versions (Go and Nuget).
> 
> > In practice, people pin versions that actually exist
> 
> > Typically, we see folks publishing higher versions of their packages as time goes by, not lower versions.
> 
> I recognize this is an edge case we probably don't need to worry about. I was just highlighting one difference of lockfiles and version pinning, as I clarified the semantics of the check.

thanks! idk who owns the naming but might be worth calling this out in a future meeting.

**should package managers provide basic support for the definition of pinned dependencies I provided earlier and what's in the docs? (i.e. exact version)?** or is the expectation to contribute to the "locked" only definition?

lock files aren't that prevalent in NuGet/.NET. this means that the .NET/NuGet ecosystem would score very low:

![image](https://github.com/user-attachments/assets/f0f77e7b-9162-46c0-ba26-4448b3433e61)

it seems like this check might benefit from partial points for exact versions and full points for locked/frozen/etc deps?

or perhaps a new check for "pinned" / "locked" package dependencies would be needed to take all this into account to separate from the CI/CD & docker checks?

Reply:
Normally, the transitive graph is resolved for you based on what the packages you consume have declared. To "pin" a transitive package to a higher version, you just add a direct dependency on it via `<PackageReference Include="SomeTransitivePackage" />`.  With CPM and transitive pinning, NuGet just does this for you during graph resolution without you having to declare the direct dependency.